models/biocro/R/write.configs.BIOCRO.R

```diff
@@ -1,92 +1,99 @@
-# are made available under the terms of the 
+# are made available under the terms of the
 # University of Illinois/NCSA Open Source License
 # which accompanies this distribution, and is available at
 # http://opensource.ncsa.illinois.edu/license.html
 #-------------------------------------------------------------------------------
 
 PREFIX_XML <- "<?xml version=\"1.0\"?>\n<!DOCTYPE config SYSTEM \"biocro.dtd\">\n"
 
-##------------------------------------------------------------------------------------------------#
 ##' convert parameters from PEcAn database default units to biocro defaults
-##' 
+##'
 ##' Performs model specific unit conversions on a list of trait values,
 ##' such as those provided to write.config
 ##' @name convert.samples.BIOCRO
 ##' @title Convert samples for biocro
 ##' @param trait.samples a matrix or dataframe of samples from the trait distribution
+##' @param biocro_version numeric, but currently only checks whether version is less than 1.0
 ##' @return matrix or dataframe with values transformed
 ##' @export
 ##' @author David LeBauer
-convert.samples.BIOCRO <- function(trait.samples) {
-  
+convert.samples.BIOCRO <- function(trait.samples, biocro_version=1.0) {
+
   if (is.list(trait.samples)) {
     trait.samples <- as.data.frame(trait.samples)
   }
-  ## first rename variables
   trait.names <- colnames(trait.samples)
-  trait.names[trait.names == "Vcmax"] <- "vmax"
-  trait.names[trait.names == "Jmax"] <- "jmax"
-  trait.names[trait.names == "leaf_respiration_rate_m2"] <- "Rd"
-  trait.names[trait.names == "cuticular_cond"] <- "b0"
-  trait.names[trait.names == "stomatal_slope.BB"] <- "b1"
-  trait.names[trait.names == "SLA"] <- "Sp"
-  trait.names[trait.names == "growth_respiration_coefficient"] <- "GrowthRespFraction"
-  trait.names[trait.names == "extinction_coefficient_diffuse"] <- "kd"
-  trait.names[trait.names == "chi_leaf"] <- "chi.l"
-  trait.names[trait.names == "quantum_efficiency"] <- "alpha"
-  
-  colnames(trait.samples) <- trait.names
-  ## Partitioning coefficients: especially leaf phenology
-  
-  ## iRhizome iStem ifrRhizome ifrStem
-  
+
   ## transform values with different units cuticular conductance - BETY default is
   ## umol; BioCro uses mol
-  if ("b0" %in% trait.names) {
-    trait.samples$b0 = udunits2::ud.convert(trait.samples$b0, "umol", "mol")
+  if ("cuticular_cond" %in% trait.names) {
+    trait.samples$cuticular_cond = udunits2::ud.convert(trait.samples$cuticular_cond, "umol", "mol")
+  }
+  if ("SLA" %in% trait.names) {
+    trait.samples$SLA = udunits2::ud.convert(trait.samples$SLA, "kg/m2", "g/cm2")
   }
-  if ("Sp" %in% trait.names) {
-    trait.samples$Sp = udunits2::ud.convert(trait.samples$Sp, "kg/m2", "g/cm2")
+
+  ## rename bety variables to match active version of biocro
+  names(trait.names) <- trait.names #looks weird, but naming this vector simplifies indexing below
+
+  name_lookup <- dplyr::tribble(
```

Comment:
@justinmcgrath similar to weather data, does the model v1 support these parameter synonyms?

I don't think that parameter synonyms are as useful for backward compatibility as met since the parameter files are fundamentally different.

Reply:
I don't think there is any cuticular conductance. Only one parameter name is accepted though. These sorts of mappings are not part of the model, or the caller, and would be better designed as its own entity, with only a specific mapping defining how the caller and model interact.

Reply:
if I understand correctly, this is the specific mapping defining how the caller and model interact. We have discussed using standalone tables in files or the database, but these tables are generally small. I was mostly wondering if there is backward compatibility between v0,9x and v1 parameter names; sounds like there is not.

Reply:
@justinmcgrath Is there a table somewhere with names and definitions of the parameters Biocro 1 recognizes? I constructed these synonyms by hand-mapping the changes between v0.9 and v1 Micanthus parameters, and could easily have missed something or gotten things wrong.

Reply:
@justinmcgrath Also, by "I don't think there is any cuticular conductance" do you mean BioCro has no input parameter for it? In that case what is b0? Do you see any others that are misidentified here?

Reply:
There's this out-of-date file: https://github.com/ebimodeling/biocro-dev/blob/master/data/biogro_parameters.r

Updating that is on my list of things to do.

I guess I would consider b0 and cuticular conductance to be different things, but it looks like many articles make no distinction.